modules/c++/xml.lite/source/Speed.cpp

```diff
@@ -0,0 +1,281 @@
+/* =========================================================================
+ * This file is part of xml.lite-c++
+ * =========================================================================
+ *
+ * (C) Copyright 2004 - 2014, MDA Information Systems LLC
+ *
+ * xml.lite-c++ is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; If not,
+ * see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <string>
+#include <vector>
+#include <map>
+#include <stdio.h>
+#include <stdlib.h>
+#include <import/str.h>
+#include <import/except.h>
+#include "xml/lite/Element.h"
+#include "xml/lite/Speed.h"
+
+//=================================================================
+void xml::lite::Speed::fromDocument(xml::lite::Document *doc, bool steal)
+{
+  setRootElement(doc->getRootElement(steal), steal);
+  buildTOC();
+}
+//=================================================================
+void xml::lite::Speed::buildTOC(const std::string &parent,
+                         xml::lite::Element const * element)
+{
+  // make a smaller, local map for bookkeeping.
+  std::map<std::string, size_t> local_registry;
+  // either both or niether parent and element must be specified
+  if (parent.empty() ^ (element == NULL))
+  {
+    throw except::Exception(
+        "xml::lite::Speed::buildTOC() called with only one of"
+        " parent path or current element."
+        " Only neither or both are allowed.");
+  }
+
+  if (element == NULL)
+  {
+    element = mRootNode;
+    // If we are starting at the top, clear out the old TOC before rebuilding.
+    mTOC.clear();
+  }
+
+  std::vector<xml::lite::Element *> childs = element->getChildren();
+  size_t sz = childs.size();
```

Comment:
You can assume this method is inlined and free, so just put it in the for loop condition